mikeio1d/result_reader_writer/result_reader_copier.py

```diff
@@ -70,26 +95,55 @@ def create_data_frame(self, data_entries, column_names):
         data_pointer_net = IntPtr(data_pointer)
         self.result_data_copier.CopyData(data_pointer_net, data_entries)
 
-        df = pd.DataFrame(data_array, index=self.time_index, columns=column_names)
+        columns = self.create_column_index(timeseries_ids, column_mode=column_mode)
+
+        df = pd.DataFrame(data_array, index=self.time_index, columns=columns)
 
         self.update_time_quantities(df)
 
         return df
 
-    def get_all_data_entries_and_column_names(self):
+    def create_column_index(
+        self,
+        timeseries_ids: List[TimeSeriesId],
+        column_mode: Optional[ResultReader.ColumnMode] = None,
+    ) -> pd.MultiIndex | pd.Index:
+        """Creates a DataFrame column from a list of TimeSeriesId objects and the current column_mode."""
+        if column_mode is None:
+            column_mode = self.column_mode
+        if column_mode == ResultReader.ColumnMode.ALL:
+            return TimeSeriesId.to_multiindex(timeseries_ids)
+        elif column_mode == ResultReader.ColumnMode.TIMESERIES:
+            return pd.Index(timeseries_ids)
+        elif column_mode == ResultReader.ColumnMode.QUERY:
+            queries = [t.to_query() for t in timeseries_ids]
+            return pd.Index([str(q) for q in queries])
+        else:
+            if column_mode in ResultReader.ColumnMode:
+                raise NotImplementedError(f"The column_mode {column_mode} is not implemented.")
+            raise ValueError(f"Unknown column_mode: {column_mode}")
+
+    def get_all_data_entries_and_timeseries_ids(self) -> Tuple[DataEntryNet, List[TimeSeriesId]]:
         data_entries = self.result_data_copier.GetEmptyDataEntriesList()
-        column_names = []
+        timeseries_ids: List[TimeSeriesId] = []
+        timeseries_ids_set = set()
         for data_set in self.data.DataSets:
             data_set = impl(data_set)
             if not self.is_data_set_included(data_set):
                 continue
 
             for data_item in data_set.DataItems:
+                data_item = impl(data_item)
                 for i in range(data_item.NumberOfElements):
                     data_entry = DataEntryNet(data_item, i)
                     data_entries.Add(data_entry)
 
-                    column_name = self.get_column_name(data_set, data_item, i)
```

Comment:
Suggestion: keep `self.get_column_name` and rename to `self.get_timeseries_id` and put the contents with the while loop in it. Make _timeseires_ids and _timeseries_ids_set "private" instance variables.